crates/fs/private/previous.rs

```diff
@@ -0,0 +1,843 @@
+
+        path.push((tail, last.to_string()));
+
+        for (dir, path_segment) in path {
+            previous_iter.path.push(PathSegmentHistory {
+                dir: Rc::clone(&dir),
+                history: PrivateNodeHistory::of(
+                    &PrivateNode::Dir(Rc::clone(&dir)),
+                    &dir.header.ratchet,
+                    discrepancy_budget,
+                    Rc::clone(&forest),
+                )?,
+                path_segment,
+            });
+        }
+
+        // For the first part of the path, we specifically set the history ourselves,
+        // because we've had `past_ratchet` passed in from the outside.
+
+        previous_iter.path[0].history.ratchets =
+            PreviousIterator::new(past_ratchet, &new_ratchet, discrepancy_budget)
+                .map_err(FsError::PreviousError)?;
+
+        Ok(previous_iter)
+    }
+
+    /// Step the history one revision back and retrieve the node at the configured path.
+    ///
+    /// Returns `None` if there is no more previous revisions.
+    pub async fn previous<B: BlockStore>(&mut self, store: &B) -> Result<Option<PrivateNode>> {
+        // Finding the previous revision of a node works by trying to get
+        // the previous revision of the path elements starting on the deepest
+        // path node working upwards, in case the history of lower nodes
+        // have been exhausted.
+        //
+        // Once another history entry on the path has been found, we proceed
+        // to work back trying to construct new history entries by going downwards
+        // on the same path from an older root revision, until we've completed
+        // the whole path and found new history entries in every segment.
+
+        if let Some(node) = self.target.previous_node(store).await? {
+            return Ok(Some(node));
+        }
+
+        let mut working_stack: Vec<(Rc<PrivateDirectory>, String)> =
+            Vec::with_capacity(self.path.len());
+
+        loop {
+            // Pop elements off the end of the path
+            if let Some(mut segment) = self.path.pop() {
+                // Try to find a path segment for which we have previous history entries
+                if let Some(prev) = segment.history.previous_dir(store).await? {
+                    segment.dir = prev;
+                    self.path.push(segment);
+                    // Once found, we can continue.
+                    break;
+                }
+
+                working_stack.push((segment.dir, segment.path_segment));
+            } else {
+                // We have exhausted all histories of all path segments.
+                // There's no way we can produce more history entries.
+                return Ok(None);
+            }
+        }
+
+        // Work downwards from the previous history entry of a path segment we found
+        for (directory, path_segment) in working_stack {
+            let ancestor = self.path.last().unwrap();
```

Comment:
How would I do that best? I don't particularly like he `unwrap` here, too. But I'm not sure how I'd get around it.
We know that `self.path` must at least have a single element, because if it didn't, then the `.pop()` call above would've returned `None` and it would've jumped out of the function.

Reply:
@matheus23 so 2 paths here, just use a ? anyway, even though we shouldn't hit this case. Or, if we really expect it to never hit, use `.expect` w/ a message when such a thing would ever panic.

Reply:
Hm. I want that "should never happen" error to pop out and stop the world when it happens, so I'd say we go for `.expect`.
I considered making it an `Err(FsError::NotFound)`, too, but that'd just hide the fact that there was something that went *really* wrong.